File: backend/models/db-functions.js
const sqlite3 = require('sqlite3');
const { connectToDatabase } = require('../config/db');

// --- INITIALISE DB ---
let db;

( async () => {
    db = await connectToDatabase();
})();


// --- PRIMITIVE TABLE FUNCTIONS ---

async function create_table(table_name, table_columns, data_types) {
    let temp = [];
    for (const [i, el] of table_columns.entries()) {
        let x = `${el} ${data_types[i]}`;
        temp.push(x);
    };

    const table_data = temp.join(', ')
    const query = `CREATE TABLE IF NOT EXISTS ${table_name} (${table_data})`;
    console.log(query);
    db.run(query);  
};

async function read_table(table_name) {
    const query = `SELECT * FROM ${table_name}`;
    const res = await db.all(query);
    return res;
}

async function insert_into_table(table_name, col, data) {
    const query = `INSERT INTO ${table_name} (${col}) VALUES (?)`;
    db.run(query, data);
};

async function update_table(table_name, col, old_val, new_val) {
    const query = `UPDATE ${table_name} SET ${col} = ? WHERE ${col} = ?`;
    db.run(query, new_val, old_val);
};

async function delete_from_table(table_name, col, val) {
    const query = `DELETE FROM ${table_name} WHERE ${col} = ?`;
    db.run(query, val);
};

async function delete_table(table_name) {
    const query = `DROP TABLE ${table_name}`;
};

// --- SETUP ---

async function create_db() {
    create_table("staff", ["staff_id", "email", "name", "password", "gender", "attendance", "role"], ["INTEGER AUTOINCREMENT", "TEXT", "TEXT", "TEXT", "TEXT", "REAL", "TEXT DEFAULT 'User'"]);
    create_table("students", ["student_id", "email", "name", "password", "year", "gender", "age", "attendance", "role"], ["INTEGER AUTOINCREMENT", "TEXT", "TEXT", "TEXT", "TEXT", "TEXT", "INTEGER", "REAL", "TEXT DEFAULT 'User"]);
    create_table("parents", ["parent_id", "email", "name", "password", "gender", "role"], ["INTEGER AUTOINCREMENT", "TEXT", "TEXT", "TEXT", "TEXT", "TEXT DEFAULT 'User"]);
    create_table("courses", ["course_id", "name"], ["INTEGER AUTOINCREMENT", "TEXT"]);
    create_table("events", ["event_id", "name", "description", "datetime", "price"], ["INTEGER AUTOINCREMENT", "TEXT", "TEXT", "TEXT", "FLOAT"]);
};

create_db()

File: backend/modules/db-functions.py
import bcrypt
import os
import sqlite3

db_path = '../instance/db.db'


# --- INITIALISATION ---

def create_db(db_path):
    create_table("students", ["id, email, name, password, year, gender, age, attendance, role"], ["INTEGER AUTOINCREMENT", "TEXT", "TEXT", "INTEGER", "TEXT", "INTEGER", "REAL", "TEXT DEFAULT 'User"])

def connect(db_path):
    return sqlite3.connect(db_path)

def db_exists(db_path):
    if not os.path(db_path):
        return False
    else:
        return True



# --- CRUD TABLE FUNCTIONS ---

def create_table(conn, table_name, col_vals, col_types):
    c = conn.cursor()

    cols_and_types = []
    for i, e in enumerate(col_vals):
        temp = f"{e} {col_types[i]}"
        cols_and_types.append(temp)

    query = f"CREATE TABLE IF NOT EXISTS {table_name} ({(', '.join(cols_and_types))})"
    c.execute(query)
    conn.commit()

def read_table(conn, table_name):
    c = conn.cursor()

    query = f"SELECT * FROM {table_name}"
    c.execute(query)
    return c.fetchall()

def update_table_add_col(conn, table_name, col_val, col_type):
    c = conn.cursor()

    query = f"ALTER TABLE {table_name} ADD COLUMN ({col_val} {col_type})"
    c.execute(query)
    conn.commit()

def update_table_remove_col(conn, table_name, col_val):
    c = conn.cursor()

    query = f"ALTER TABLE {table_name} DROP COLUMN {col_val}"
    c.execute(query)
    conn.commit()

def delete_table(conn, table_name):
    c = conn.cursor()

    query = f"DROP TABLE {table_name}"
    c.execute(query)
    conn.commit()


# --- CRUD TABLE DATA FUNCTIONS ---

def add_to_table(conn, table_name, data_vals):
    c = conn.cursor()

    query = f"INSERT INTO {table_name} ({(', ').join(data_vals)})"
    c.execute(query)
    conn.commit()

def edit_table_data(conn, table_name, col, data, id):
    c = conn.cursor()

    query = f"UPDATE {table_name} SET {col} = {data} WHERE id = {id}"
    c.execute(query)
    conn.commit()

def remove_from_table(conn, table_name, id):
    c = conn.cursor()

    query = f"DELETE FROM {table_name} WHERE id = {id}"
    c.execute(query)
    conn.commit()


# --- USER FUNCTIONS ---

def create_user(conn, table_name, data):
    pw = data[2]

    bytes = pw.encode('utf-8')
    salt = bcrypt.gensalt(10)
    hash_pw = bcrypt.hashpw(bytes, salt)

    data[2] = hash_pw

    add_to_table(conn, table_name, data)